cli/printlog/compact/compact.go

```diff
@@ -0,0 +1,411 @@
+
+	var spawns []*spb.SpawnExec
+	for {
+		s, err := slr.GetSpawnExec(sort)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return fmt.Errorf("failed getting spawn exec: %s", err)
+		}
+		if sort {
+			spawns = append(spawns, s)
+			continue
+		}
+
+		if err := printSpawnExec(s); err != nil {
+			return err
+		}
+	}
+	if !sort {
+		return nil
+	}
+
+	// TODO(sluongng): move StableSort to separate func and write tests
+
+	outputToSpawns := make(map[string][]*spb.SpawnExec)
+	for _, s := range spawns {
+		for _, o := range s.GetActualOutputs() {
+			ss, ok := outputToSpawns[o.GetPath()]
+			if !ok {
+				outputToSpawns[o.GetPath()] = []*spb.SpawnExec{s}
+				continue
+			}
+			outputToSpawns[o.GetPath()] = append(ss, s)
+		}
+	}
+
+	blockedBy := newIdentitySetMultiMap[*spb.SpawnExec, *spb.SpawnExec]()
+	blocking := newIdentitySetMultiMap[*spb.SpawnExec, *spb.SpawnExec]()
+
+	queue := make(priorityQueue, len(spawns))
+
+	for _, s := range spawns {
+		if s == nil {
+			// TODO: why nil?
+			continue
+		}
+		blocked := false
+		for _, input := range s.GetInputs() {
+			for _, blocker := range outputToSpawns[input.GetPath()] {
+				blockedBy.put(s, blocker)
+				blocking.put(blocker, s)
+				blocked = true
+			}
+		}
+		if blocked {
+			heap.Push(&queue, s)
+		}
+	}
+
+	for queue.Len() > 0 {
+		s := heap.Pop(&queue).(*spb.SpawnExec)
+		if s == nil {
+			// TODO: why nil?
+			continue
+		}
+		if err := printSpawnExec(s); err != nil {
+			return err
+		}
+
+		for blocked := range blocking.get(s) {
+			if blocked == nil {
+				// TODO: why nil?
+				continue
+			}
+			blockedBy.remove(blocked, s)
+			if !blockedBy.contains(blocked) {
+				heap.Push(&queue, blocked)
+			}
+		}
+	}
+
+	return nil
+}
+
+type priorityQueue []*spb.SpawnExec
+
+func (q priorityQueue) Len() int {
+	return len(q)
+}
+
+func (q priorityQueue) Less(i, j int) bool {
+	return toCompareString(q[i]) > toCompareString(q[j])
+}
+
+func toCompareString(s *spb.SpawnExec) string {
+	// Sort by comparing the path of the first output. We don't want the sorting to
+	// rely on file hashes because we want the same action graph to be sorted in the
+	// same way regardless of file contents.
+	if len(s.GetListedOutputs()) > 0 {
+		return "1_" + s.GetListedOutputs()[0]
+	}
+
+	// Get a proto with only stable information from this proto
+	stripped := &spb.SpawnExec{
+		CommandArgs:          s.GetCommandArgs(),
+		EnvironmentVariables: s.GetEnvironmentVariables(),
+		Platform:             s.GetPlatform(),
+		Inputs:               s.GetInputs(),
+		Mnemonic:             s.GetMnemonic(),
+	}
+	return "2_" + stripped.String()
+}
+
+func (q priorityQueue) Swap(i, j int) {
+	q[i], q[j] = q[j], q[i]
+}
+
+func (q *priorityQueue) Push(s any) {
+	*q = append(*q, s.(*spb.SpawnExec))
+}
+
+func (q *priorityQueue) Pop() any {
+	old := *q
+	n := len(old)
+	s := old[n-1]
+	old[n-1] = nil
+	*q = old[:n-1]
+	return s
+}
+
+type identitySetMultiMap[T, V comparable] struct {
```

Comment:
Compared to the Java implementation, this doesn't enforce reference equality semantics for T and V, so I guess it's just a `setMultiMap` at this point.

Reply:
I converted this to non-generic